keps/sig-node/4960-container-stop-signals/README.md

```diff
@@ -0,0 +1,382 @@
+
+```go
+func (m *kubeGenericRuntimeManager) killContainer(ctx context.Context, pod *v1.Pod, containerID kubecontainer.ContainerID, containerName string, message string, reason containerKillReason, gracePeriodOverride *int64, ordering *terminationOrdering) error {
+// ...
+	err := m.runtimeService.StopContainer(ctx, containerID.ID, gracePeriod, containerSpec.StopSignal)
+```
+
+Once the StopSignal from the ContainerSpec is passed down to the container runtime via the updated `StopContainerRequest`, we can use the value of the StopSignal from the container runtime's implementation of `stopContainer` method. In the case of containerd, the changes required would look like the following:
+
+```diff
+//internal/cri/server/container_stop.go
+
+func (c *criService) StopContainer(ctx context.Context, r *runtime.StopContainerRequest) (*runtime.StopContainerResponse, error) {
+// ...
+-	if err := c.stopContainer(ctx, container, time.Duration(r.GetTimeout())*time.Second); err != nil {
++ 	if err := c.stopContainer(ctx, container, time.Duration(r.GetTimeout())*time.Second, r.GetStopSignal()); err != nil {
+		return nil, err
+	}
+// ...
+}
+
+-func (c *criService) stopContainer(ctx context.Context, container containerstore.Container, timeout time.Duration) error {
++func (c *criService) stopContainer(ctx context.Context, container containerstore.Container, timeout time.Duration, containerStopSignal string) error {
+//...
+    if timeout > 0 {
+	    stopSignal := "SIGTERM"
+-       if container.StopSignal != "" {
++       if containerStopSignal != "" {
++   	    stopSignal = containerStopSignal 
++       } else if  container.StopSignal != "" {
+			stopSignal = container.StopSignal
+		} else {
+// rest of the code...
+```
+
+Obviously these are not all of the required changes, but a quick idea of what would need to be done. Find the entire diff for containerd [here](https://github.com/containerd/containerd/compare/main...sreeram-venkitesh:containerd:added-custom-stop-signal?expand=1).
+### User Stories (Optional)
+
+#### Story 1
+
+Kubernetes by default sends a SIGTERM to all containers while killing them. When running nginx on Kubernetes, this can result in nginx dropping requests as reported [here](https://github.com/Kong/kubernetes-ingress-controller/pull/283). The current solution for this issue would be to build custom images with a SIGQUIT stop signal or to write a PreStop lifecycle hook that manually kills the process gracefully, which is what is done in the PR. If we had stop signal support at the Container spec level, this would've been easier and straightforward to implement. Users wouldn't have to patch the applications running on Kubernetes to handle different termination behavior. This is also similar to [this issue](https://github.com/github/resque/pull/21). 
+
+### Risks and Mitigations
+
+I don't see any issues with adding a new field for the stop signal because this is an optional feature that users.
```

Comment:
```suggestion
I don't see any issues with adding a new field for the stop signal because this is an optional feature.
```

Reply:
A risk is adding the complexity of signal handling to pod/container spec users unhandled signals, hung pods, erratic results..  I don't see kubectl options in the kep, was that considered.

Reply:
Understood the risks. I don't get what you meant with kubectl options. Are you talking about adding stop signal as an option to kubectl run maybe? Kevin mentioned adding stop signal as an option for `crictl stop`, which I've added to the KEP.